src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker_Generators.cs

```diff
@@ -0,0 +1,159 @@
+                solution,
+                this.ProjectState.Id,
+                (service, solutionChecksum, cancellationToken) => service.GetContentsAsync(
+                    solutionChecksum, projectId, documentsToAddOrUpdate.ToImmutable(), cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!generatedSourcesOpt.HasValue)
+                return null;
+
+            var generatedSources = generatedSourcesOpt.Value;
+            Contract.ThrowIfTrue(generatedSources.Length != documentsToAddOrUpdate.Count);
+
+            // Now go through and produce the new document states, using what we have already if it is unchanged, or
+            // what we have retrieved for anything new/changed.
+            using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
+            foreach (var (documentIdentity, contentIdentity) in infos)
+            {
+                var addOrUpdateIndex = documentsToAddOrUpdate.IndexOf(documentIdentity.DocumentId);
+                if (addOrUpdateIndex >= 0)
+                {
+                    // a document whose content we fetched from the remote side.
+                    var generatedSource = generatedSources[addOrUpdateIndex];
+                    var sourceText = SourceText.From(
+                        generatedSource, contentIdentity.EncodingName is null ? null : Encoding.GetEncoding(contentIdentity.EncodingName), contentIdentity.ChecksumAlgorithm);
+
+                    var generatedDocument = SourceGeneratedDocumentState.Create(
+                        documentIdentity,
+                        sourceText,
+                        ProjectState.ParseOptions!,
+                        ProjectState.LanguageServices);
+                    Contract.ThrowIfTrue(generatedDocument.GetTextChecksum() != contentIdentity.Checksum, "Checksums must match!");
+                    generatedDocumentsBuilder.Add(generatedDocument);
+                }
+                else
+                {
+                    // a document that already matched something locally.
+                    var existingDocument = generatorInfo.Documents.GetRequiredState(documentIdentity.DocumentId);
+                    Contract.ThrowIfTrue(existingDocument.Identity != documentIdentity, "Identies must match!");
+                    Contract.ThrowIfTrue(existingDocument.GetTextChecksum() != contentIdentity.Checksum, "Checksums must match!");
+                    generatedDocumentsBuilder.Add(existingDocument);
+                }
+            }
+
+            var generatedDocuments = new TextDocumentStates<SourceGeneratedDocumentState>(generatedDocumentsBuilder.ToImmutableAndClear());
+            var compilationWithGeneratedFiles = compilationWithoutGeneratedFiles.AddSyntaxTrees(
+                await generatedDocuments.States.Values.SelectAsArrayAsync(state => state.GetSyntaxTreeAsync(cancellationToken)).ConfigureAwait(false));
```

Comment:
>state [](http://example.com/codeflow?start=74&length=5)

Isn't there an override that takes in the cancellationtoken too so that this can be static?

Reply:
yup.  will switch.  would be nice to have an analyzer for that :)

Reply:
done!